### PR TITLE
feat: [M12] LSP oracle service (multi-op: definition/references/completions/quickinfo)

### LINKED ISSUE #220: [M12] LSP oracle service (multi-op: definition/references/completions/quickinfo)
Part of #198 (SSI-P0-T1, survives the MHM pivot). Extend the diagnostics-only oracle (`src/lsp/ts_lsp.py`) to a multi-op LSP client: definition / references / completions / quickinfo / documentSymbols, warm program, incremental versioning + a throughput SLA. Consumers: the code eval suite (type-aware/cross-file) and completion-list logit masking. #211 is the diagnostic-settle sub-fix.

Accept: median warm `diagnostics` <50ms on a 5k-file project; ≥200 calls/s; **kill: <~20 calls/s** ⇒ rescope decode-time masking to offline.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #220

## Summary
- Add `TsLspService` (`src/lsp/ts_service.py`) -- a NEW warm, multi-op `typescript-language-server` client alongside the untouched, one-shot `TsLspOracle` (`src/lsp/ts_lsp.py`) and `CompositeOracle` (`src/lsp/oracle.py`, both zero-diff). Stable URIs, monotonically versioned `didChange` edits, lazy `didOpen`, and `definition`/`references`/`completions`/`quickinfo`/`document_symbols` in addition to version-aware first-push-wins `diagnostics` (no settle window, per #211's confirmed single-push completeness on the pinned 5.3.0 server).
- `src/lsp/diagnostics.py`: add `offset_to_lsp_position` (+ `_char_index_to_utf16_col`), the LSP-position inverse of `line_col_to_offset`, needed so every op can be addressed by character offset.
- `scripts/bench_ts_lsp.py`: **new** SLA harness -- a deterministic N-file synthetic project generator, a BLIND-guarded sanity probe (known-answer checks on every op before any timing starts), and per-op median/mean/min/p95 latency + calls/s measurement.
- `scripts/eval_code_suite.py:359`: fix the live `oracle.diagnose(artifact)` -> `oracle.diagnostics(artifact)` `AttributeError` bug in `_run_tsc` (the only `.diagnose(` call site in the repo -- `--suites tsc` raised on its first row), with a regression test (`tests/test_eval_code_suite.py`) pinning the real method name via a stub oracle with no `diagnose` method.
- `tests/test_ts_service_mechanism.py`: **new**, binary-free `os.pipe()` scripted-server tests -- the actual CI gate for `TsLspService` (verified green with `PATH=/usr/bin:/bin`, no node toolchain).
- `tests/test_ts_service.py`: **new**, live `typescript-language-server` tests (cross-file definition/references fan-in, restart recovery, no-orphan-process, parity with `TsLspOracle` on a #194 record). Skips cleanly without the toolchain.
- `tests/test_lsp_diagnostics.py`: round-trip + edge-case tests for `offset_to_lsp_position`.
- `tests/test_import_guard.py`: `src.lsp.ts_service` added to `PORTABLE_MODULES`.

## Measured SLA result (real numbers, not tuned to pass)
`scripts/bench_ts_lsp.py --n-files 5000 --warmup 20 --iters 200` (seed 0), sanity probe passed on every check:

| op | median | calls/s |
|---|---:|---:|
| `diagnostics` (edit->re-check, the #226 shape) | 378.1 ms | 3.1 |
| `completions` | 0.38 ms | **2,629** |
| `definition` | 1.74 ms | 595 |
| `references` | 506.0 ms | 2.0 |
| `quickinfo` | 0.31 ms | 3,102 |
| `document_symbols` | 0.42 ms | 2,332 |

**Verdict: WARN.** The throughput gate (`completions` >= 200 calls/s, #226's decode-time masking) clears comfortably at 13x the bar. The latency gate (`diagnostics` median < 50ms, the type-aware eval arm) misses by ~8x at 5000 files -- a real, measured finding recorded in `docs/design/12-lsp-in-the-loop.md` and `results/ts_lsp_bench.json`, not a harness tuned until it passed. `n_restarts=0`, `n_timeouts=0`, 100% non-empty rate on every op except the (expected-clean) benign-edit `diagnostics` loop -- the sanity probe rules out a BLIND/empty-and-fast measurement.

## Testing
- [x] Tests added/updated
- [x] All tests passing (`pytest -q -rs`: 1327 passed, 13 skipped -- all pre-existing/unrelated)
- [x] `tests/test_import_guard.py` green with `src.lsp.ts_service` listed
- [x] `git diff --stat` confirms zero diff on `src/lsp/ts_lsp.py` and `src/lsp/oracle.py`
- [x] `scripts/eval_code_suite.py --suites tsc` completes instead of raising `AttributeError`
- [x] Mechanism tests verified green with no node toolchain on `PATH` (the real CI gate)
- [x] Live tests (`tests/test_ts_service.py`, 10 tests) pass against the pinned local `typescript-language-server` 5.3.0
- [x] `scripts/bench_ts_lsp.py` produced a real verdict (WARN, not BLIND)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_019K3gzzSsE7RJjyNp4QUQff

### CODE REVIEW

File: src/lsp/diagnostics.py [hunk 1]
```diff
@@ -178,6 +179,39 @@ def line_col_to_offset(source: str, line: int, col: int) -> int:
...
+    (EOF) is legal and must work -- it is the decode-frontier position #226 uses.
+    """
+    if offset < 0 or offset > len(source):
+        raise ValueError(f"offset {offset} out of range for source of length {len(source)}")
+    starts = _line_start_offsets(source)
+    line0 = bisect.bisect_right(starts, offset) - 1
+    line_text = source.split("\n")[line0]
+    char_index = offset - starts[line0]
+    character0 = _char_index_to_utf16_col(line_text, char_index) - 1
+    return {"line": line0, "character": character0}
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/lsp/diagnostics.py:223]: ✅ Fixed in db76c79dc6249bf01157753ccd8008140b1eb189 — offset_to_lsp_position/line_col_to_offset now slice the needed line from the already-computed line-start offsets instead of `source.split("\\n")`, avoiding the O(len(source)) full-list allocation on every call.

File: src/lsp/ts_service.py [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+            return
+        uri = self._uri(path)
+        text = self._files[path]
+        with self._diag_lock:
+            self._update_marker.setdefault(uri, self._diag_seq_counter)
+        self._endpoint.notify("textDocument/didOpen", {
+            "textDocument": {"uri": uri, "languageId": "typescript",
+                              "version": self._versions[path], "text": text},
+        })
+        self._open.add(path)
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/lsp/ts_service.py]: ✅ Fixed in db76c79dc6249bf01157753ccd8008140b1eb189 — added a _notify_retrying helper: _ensure_open's didOpen notify and update()'s didChange notify now retry once after _restart() on OSError (e.g. BrokenPipe from the server dying between _ensure_alive()'s poll() and the write), preserving diagnostics()'s documented never-raises-on-a-dead-server contract. (_timed_request's own path was already covered — ConnectionError, which it already catches, is a superclass of BrokenPipeError in the stdlib, so no change was needed there.)

File: scripts/bench_ts_lsp.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    # diagnostics: an EDIT -> RE-CHECK cycle, timed TOGETHER -- the shape
+    # #226's consumer actually has (one edit per generated token). Timing a
+    # cache hit and calling it a re-check would be the same BLIND failure in
+    # a different coat.
+    counter = [0]
+
+    def _edit_and_check(path):
+        counter[0] += 1
+        service.update(path, files[path] + f"\n// bench edit {counter[0]}\n")
+        return service.diagnostics(path)
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply scripts/bench_ts_lsp.py:279]: ⏭️ Reviewed - No changes made. files[path] is the fixed base dict from _generate_project and is never reassigned anywhere in _run_measurements/_measure_op, so each _edit_and_check iteration rebuilds text as files[path] + suffix from the SAME unchanged base every time, not from the previous iteration's edited text. The file does not grow across iterations (only the counter's digit count grows, which is negligible), so there's no accumulating-size bias to fix here.

**Comment by travisgalloway** [comment]
Addressed review feedback:

- @copilot-pull-request-reviewer: overview only, no separate actionable content beyond the 3 inline comments below — all 3 resolved (2 fixed in db76c79, 1 reviewed as not-applicable).
